automates/text_reading/webapp/app/controllers/HomeController.scala

```diff
@@ -219,10 +261,17 @@ class HomeController @Inject()(cc: ControllerComponents) extends AbstractControl
       json("arguments").obj("maxSVOgroundingsPerVar").num.toInt
     } else maxSVOgroundingsPerVarDefault
 
+    val textInputFormat = if (jsonKeys.contains("arguments")) {
+      val args = json("arguments")
+      if (args.obj.keys.toList.contains("text_input_format")) {
```

Comment:
@maxaalexeeva you should be able to do the following for the key membership check on line 266:
```scala
args.obj.contains("text_input_format")
```

This method should attempt to hash the string `text_input_format` into the `args` object. That is an _O(1)_ operation, whereas (if I am understanding this correctly) then your method is checking if the string `text_input_format` is in the list of keys that are a part of the object and that is an _O(n)_ operation. Efficiency probably isn't too big of a deal here, but we should always strive for _O(1)_ operations when the opportunity presents itself.

Try out the method I have above and see if it works (I'm not actually sure if the syntax is correct). I found this [link](https://alvinalexander.com/scala/how-to-text-existence-key-value-map-scala-cookbook/) that seems to think we can use that syntax. If it doesn't work then I'll search for another way to do an _O(1)_ check.

Reply:
@pauldhein, nice! Didn't know of that way. I think it worked. Also, I use that site you linked a lot! Great explanations!

Reply:
Glad it worked. Thank you for making the change!